lab1/myswitch_fifo.py

```diff
@@ -0,0 +1,77 @@
+from switchyard.lib.userlib import *
+
+
+class ForwaridngTable:
+    def __init__(self, size=5):
+        self.array = [None for i in range(size)]
+        self.pos = 0
+        self.size = size
+        self.map = {}
+        return
+
+    def __iter__(self):
+        return iter(self.map)
+
+    def __getitem__(self, item):
+        return self.map[item]
+
+    def update(self, addr, port):
+        """ if addr is already in map, we set it again
+        without changing anything else, (as the port may
+        have changed).
+        If addr not in map, we update ForwaridngTable in
+        a fifo manner, e.g. if we have a vacant spot, we
+        will just add the new addr/port pair; otherwise,
+        if we do not have a vacancy, we'll need to evict
+        the oldest item from the ForwaridngTable, then add
+        the new item
+        """
+        if addr in self.map:
+            self.map[addr] = port
+            return
+
+        self.map[addr] = port
+        if self.array[self.pos] == None:
+            self.array[self.pos] = addr
+        else:
+            evict = self.array[self.pos]
+            del self.map[evict]
+            self.array[self.pos] = addr
+
+        self.pos = (self.pos + 1) % self.size
+
+        return
+
+
+def main(net):
+    my_interfaces = net.interfaces()
+    mymacs = [intf.ethaddr for intf in my_interfaces]
+    forwarding_table = ForwaridngTable()
+
+    while True:
+        try:
+            timestamp, input_port, packet = net.recv_packet()
+        except NoPackets:
+            log_debug("No packets received!")
+            continue
+        except Shutdown:
+            log_debug("Received signal for shutdown!")
+            return
+
+        forwarding_table.update(packet[0].src, input_port)
+
+        log_debug("In {} received packet {} on {}".format(
+            net.name, packet, input_port))
+        if packet[0].dst in mymacs:
+            log_debug("Packet intended for me")
+        else:
+            if packet[0].dst in forwarding_table:
+                net.send_packet(forwarding_table[packet[0].dst], packet)
+            else:
+                for intf in my_interfaces:
+                    if input_port != intf.name:
+                        log_debug("Flooding packet {} to {}".format(
+                            packet, intf.name))
+                        net.send_packet(intf.name, packet)
```

Comment:
It's unlikely, but for completeness shall we add a ValueError exception handler?

Reply:
Yes, we should handle that. I've pushed updates.